public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx

```diff
@@ -1,45 +1,107 @@
 import { LokiDatasource } from '../datasource';
 import { LokiQuery } from '../types';
 import { LokiQueryField } from './LokiQueryField';
 
 type Props = QueryEditorProps<LokiDatasource, LokiQuery>;
 
-export const LokiQueryEditor = memo(function LokiQueryEditor(props: Props) {
-  const { query, data, datasource, onChange, onRunQuery } = props;
+interface State {
+  legendFormat: string;
+}
 
-  let absolute: AbsoluteTimeRange;
+export class LokiQueryEditor extends PureComponent<Props, State> {
+  // Query target to be modified and used for queries
+  query: LokiQuery;
 
-  if (data && data.request) {
-    const { range } = data.request;
-    absolute = {
-      from: range.from.valueOf(),
-      to: range.to.valueOf(),
-    };
-  } else {
-    absolute = {
-      from: Date.now() - 10000,
-      to: Date.now(),
+  absolute: AbsoluteTimeRange;
+
+  constructor(props: Props) {
+    super(props);
+    // Use default query to prevent undefined input values
+    const defaultQuery: Partial<LokiQuery> = { expr: '', legendFormat: '' };
+    const query = Object.assign({}, defaultQuery, props.query);
+    this.query = query;
+    // Query target properties that are fully controlled inputs
+    this.state = {
+      // Fully controlled text inputs
+      legendFormat: query.legendFormat,
     };
+
+    const { data } = this.props;
+
+    if (data && data.request) {
+      const { range } = data.request;
+      this.absolute = {
+        from: range.from.valueOf(),
+        to: range.to.valueOf(),
+      };
+    } else {
+      this.absolute = {
+        from: Date.now() - 10000,
+        to: Date.now(),
+      };
+    }
   }
 
-  return (
-    <div>
-      <LokiQueryField
-        datasource={datasource}
-        query={query}
-        onChange={onChange}
-        onRunQuery={onRunQuery}
-        history={[]}
-        data={data}
-        absoluteRange={absolute}
-      />
-    </div>
-  );
-});
+  onFieldChange = (query: LokiQuery, override?: any) => {
+    this.query.expr = query.expr;
+  };
+
+  onLegendChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
+    const legendFormat = e.currentTarget.value;
+    this.query.legendFormat = legendFormat;
+    this.setState({ legendFormat });
+  };
+
+  onRunQuery = () => {
+    const { query } = this;
+    this.props.onChange(query);
+    this.props.onRunQuery();
+  };
+
+  render() {
+    const { datasource, query, data } = this.props;
+    const { legendFormat } = this.state;
+
+    return (
+      <div>
+        <LokiQueryField
+          datasource={datasource}
+          query={query}
+          onChange={this.onFieldChange}
+          onRunQuery={this.onRunQuery}
+          history={[]}
+          data={data}
+          absoluteRange={this.absolute}
+        />
+
+        <div className="gf-form-inline">
+          <div className="gf-form">
+            <InlineFormLabel
+              width={7}
+              tooltip="Controls the name of the time series, using name or pattern. For example
```

Comment:
This should mention that it only applies to queries that return metrics.

Reply:
Mentioned it as such: "The legend only applies to metric queries."

Is this wording fine with you?